stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -490,10 +503,85 @@ impl NakamotoBlockHeader {
     }
 
     /// Verify the block header against an aggregate public key
-    pub fn verify_signer(&self, signer_aggregate: &Point) -> bool {
-        let schnorr_signature = &self.signer_signature.0;
+    pub fn verify_threshold_signer(
+        &self,
+        signer_aggregate: &Point,
+        signature: &ThresholdSignature,
+    ) -> bool {
         let message = self.signer_signature_hash().0;
-        schnorr_signature.verify(signer_aggregate, &message)
+        signature.verify(signer_aggregate, &message)
+    }
+
+    /// Verify the block header against the list of signer signatures
+    ///
+    /// Validate against:
+    /// - Any invalid signatures (eg not recoverable or not from a signer)
+    /// - Any duplicate signatures
+    /// - At least the minimum number of signatures (based on total signer weight
+    /// and a 70% threshold)
+    /// - Order of signatures is maintained vs signer set
+    pub fn verify_signer_signatures(&self, reward_set: &RewardSet) -> Result<(), ChainstateError> {
+        let message = self.signer_signature_hash();
+        let Some(signers) = &reward_set.signers else {
+            return Err(ChainstateError::InvalidStacksBlock(
+                "No signers in the reward set".to_string(),
+            ));
+        };
+
+        let mut total_weight_signed: u32 = 0;
+        // `last_index` is used to prevent out-of-order signatures
+        let mut last_index = None;
+
+        let total_weight = signers.iter().map(|s| s.weight).sum::<u32>();
+
+        // HashMap of <PublicKey, (Signer, Index)>
+        let signers_by_pk = signers
+            .iter()
+            .enumerate()
+            .map(|(i, signer)| (signer.signing_key, (signer.clone(), i)))
+            .collect::<HashMap<_, _>>();
+
+        for signature in &self.signer_signature {
+            let public_key = Secp256k1PublicKey::recover_to_pubkey(message.bits(), signature)
+                .map_err(|_| {
+                    ChainstateError::InvalidStacksBlock(format!(
+                        "Unable to recover public key from signature {}",
+                        signature.to_hex()
+                    ))
+                })?;
+
+            let (signer, signer_index) = signers_by_pk
+                .get(public_key.to_bytes().as_slice())
+                .ok_or_else(|| {
+                    ChainstateError::InvalidStacksBlock(format!(
+                        "Public key {} not found in the reward set",
+                        public_key.to_hex()
+                    ))
+                })?;
+
+            // Enforce order of signatures
+            match last_index {
```

Comment:
Style nit: if there's at most two match cases, we use `if let` notation like so:

```rust
If let Some(index) = last_index.as_ref() {
   if *index >= *signer_index {
      return Err(ChainstateError::InvalidStacksBlock("Signatures are out of order".to_string())));
   }
} else {
   last_index = Some(*signer_index);
}
```